### ISSUE TITLE
[heft] Unable to separate invocations for unit tests, coverage collection, and integration tests

### ISSUE THREAD

**Issue by bryandrenner** [microsoft/rushstack#3073]
## Summary

When I invoke standard Jest, I'm able to distinguish between three test-running scenarios:

- **Running all unit tests, collecting code coverage metrics.** It is useful to distinguish this scenario, because it gives me fast and holistic feedback without depending on integrated external services.
- **Selectively running unit tests in the file that I'm currently authoring.** It is useful to distinguish this scenario, because it gives me very fast feedback about the portion of implementation that I'm currently focusing on.
- **Running slower integration tests.** It is useful to distinguish this scenario, because integration tests have a dedicated global setup and teardown, they don't inform unit test coverage metrics, they are generally slow, and they require distinct setup and connectivity.

Heft-Jest doesn't appear to support these scenarios, because it only supports a single Jest config file, and it does not support Jest's standard variety of CLI arguments (such as `--coverage`). This necessitates cumbersome workflows for authoring and running tests. For example, if coverage is enabled in my configuration, then I get code coverage metrics and warnings when I only wanted to run tests from a single file. If I provide a global setup and teardown, then slow, integration-oriented code executes even for unit tests.

## Details

Heft-Jest should support more standard Jest CLI options, especially `--coverage`.

I know Heft's design philosophy is to support only a single config file, so perhaps Heft-Jest could provide its own global-setup-and-teardown strategy.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@rushstack/heft` version? | ^0.42.2 |
| Operating system? | Mac |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 14.18.1 |

**Comment by D4N14L** [comment]
Could you provide a list of args that you believe should be funneled through? Other than the `--coverage` arg, I'm guessing the `--config` arg would probably be something that you'd want.

Originally the `--config` arg was able to be provided via the plugin options in `heft.json`, but this was removed after discussion that this path really shouldn't need to be changed for the standard scenario. However, I can see a valid use-case for multiple configs supporting different scenarios that can be pointed to using `heft test --config ...`. @pgonzal thoughts?

As a step toward this goal (in PR #3076), I've moved the CLI args out of Heft and into the plugin itself, and added a `--config` arg. This arg may not make it through PR depending on what Pete has to say, but I do generally agree that the `--config` arg would be useful in this specific case (and not specifying it will default back to `config/jest.config.json`)

**Comment by bryandrenner** [comment]
`--coverage` and its alias, `--collectCoverage`, would be useful for triggering code coverage collection.

`--setupFilesAfterEnv` would be useful for initializing dependencies for integration tests.

**Comment by Faithfinder** [comment]
I'm in general interested in the approach to various jest setup files in `heft` world. Maybe it just needs some documenting, but:

I would love writing the setup files in typescript as well, but how would you specify them? Jest can't run typescript, obviously, so I need to reference the compiled file somehow. But the jest config is outside of `src` folder, so what, do I need to reference `../lib/jest.setup.js` in the config while actually creating `../src/jest.setup.ts` or something? And then again, the setup file isn't source code per se, I think they are often enough kept outside of `src` altogether...

**Comment by D4N14L** [comment]
@bryandrenner thanks for the feedback. Will look into making a PR after the Jest update goes in.

@Faithfinder you have a couple options in that case. If you want to create your setup file in TS, then yes you would need to reference it by specifying `../lib/path/to/setup/file.js`. This is actually how we include the `jest-global-setup.js` file from the plugin, as well as our custom resolver: https://github.com/microsoft/rushstack/blob/e0ad66bfb134450804be5476c179894f21d5b524/heft-plugins/heft-jest-plugin/includes/jest-shared.config.json#L59

If you want to keep it out of your `src` folder, then you'd need to craft the setup file in JS and reference it that way.

**Comment by Faithfinder** [comment]
> If you want to keep it out of your `src` folder, then you'd need to craft the setup file in JS and reference it that way.

I see. It's not a huge deal, I suppose, but the ergonomics are somewhat worse than the "standard" experience with transformers since Jest applies transformers to setup files as well. I guess I'll go create a separate issue? Even if the consensus will be that it's not worth the team's time, the issue could be linked to, because I'm probably not the last person to think about this

**Comment by D4N14L** [comment]
Yeah, we intentionally avoid the use of TypeScript transpiling transformers due to the improved performance provided by transpiling first then running the transpiled code through Jest. While I think the issue will be resolved due to the by-design nature, it makes sense to have an issue to track.
